eos/utils/expression-visitors.cc

```diff
@@ -0,0 +1,338 @@
+                }
+
+                // Make observable
+                auto observable = Observable::make(
+                    e.observable->name(),
+                    this->_parameters,
+                    this->_kinematics,
+                    this->_options
+                    );
+                // Clear alias map
+                _kinematics.clear_aliases();
+
+                return ObservableExpression(observable, e.kinematics_specification);
+            }
+    };
+
+    // Copies one expression tree to a new tree, replacing ObservableNameExpression objects with ObservableExpression objects
+    class ExpressionMaker
+    {
+        private:
+            Parameters      _parameters;
+            Kinematics      _kinematics;
+            Options         _options;
+            ParameterUser * _parameter_user;
+
+        public:
+            ExpressionMaker(const Parameters & parameters, const Kinematics & kinematics, const Options & options, ParameterUser * parameter_user = nullptr) :
+                _parameters(parameters),
+                _kinematics(kinematics),
+                _options(options),
+                _parameter_user(parameter_user)
+            {
+            }
+
+            Expression visit(const BinaryExpression & e)
+            {
+                return BinaryExpression(e.op, e.lhs.accept_returning<Expression>(*this), e.rhs.accept_returning<Expression>(*this));
+            }
+
+            Expression visit(const ConstantExpression & e)
+            {
+                return ConstantExpression(e);
+            }
+
+            Expression visit(const ObservableNameExpression & e)
+            {
+                const auto & kinematics_values = e.kinematics_specification.values;
+                const auto & kinematics_aliases = e.kinematics_specification.aliases;
+
+                // Set or alias kinematic specifications
+                for (const auto & value : kinematics_values)
+                {
+                    _kinematics.declare(value.first, value.second);
+                }
+                for (const auto & alias : kinematics_aliases)
+                {
+                    _kinematics.alias(alias.first, alias.second);
+                }
+
+                // Make observable
+                auto observable = Observable::make(
+                    e.observable_name,
+                    this->_parameters,
+                    this->_kinematics,
+                    this->_options
+                    );
+                // Clear alias map
+                _kinematics.clear_aliases();
+
+                // Record the used parameters
+                if (_parameter_user)
+                {
+                    _parameter_user->uses(*observable);
+                }
+
+                return ObservableExpression(observable, e.kinematics_specification);
+            }
+
+            Expression visit(const ObservableExpression & e)
+            {
+                return ObservableExpression(e);
```

Comment:
Ahh, bug! The return value will not necessarily be bound to the same Parameters (``_parameters``). I think we need to clone the existing observable, using ``_parameters`` as the argument to clone. Maybe you were right in the first place that this case should return an error...